src/ua_types_encoding_binary.c

```diff
@@ -1345,6 +1347,115 @@ encodeBinaryStruct(const void *src, const UA_DataType *type, Ctx *ctx) {
+                optFieldIterationPtr += sizeof(size_t);
+            if(*(void* const*)optFieldIterationPtr != NULL){
+                encodingMask |= (UA_UInt32) 1 << optFieldCounter;
+            }
+            optFieldIterationPtr += sizeof(void *);
+            optFieldCounter++;
+        } else {
+            if(m->isArray)
+                optFieldIterationPtr += sizeof(size_t);
+            optFieldIterationPtr += mt->memSize;
+        }
+    }
+    ret = UInt32_encodeBinary(&encodingMask, &UA_TYPES[UA_TYPES_UINT32], ctx);
+
+    /* Loop over members */
+    for(size_t i = 0, o = 0; i < membersSize; ++i) {
+        const UA_DataTypeMember *m = &type->members[i];
+        const UA_DataType *mt = &typelists[!m->namespaceZero][m->memberTypeIndex];
+        ptr += m->padding;
+
+        if(m->isOptional){
+            /* This optional field is not contained */
+            if(!(encodingMask & (UA_UInt32) ( (UA_UInt32) 1<<(o++)))){
+                if(m->isArray)
+                    ptr += sizeof(size_t);
+                ptr += sizeof(void *);
+            } else { //this field is contained
+                if(m->isArray){
+                    const size_t length = *((const size_t *) ptr);
+                    ptr += sizeof(size_t);
+                    ret = Array_encodeBinary(*(void *UA_RESTRICT const *) ptr, length, mt, ctx);
+                    ptr += sizeof(void *);
+                } else {
+                    ret = encodeWithExchangeBuffer(*(void* const*) ptr, mt, ctx);
+                    ptr += sizeof(void *);
+                }
+            }
+        } else {
+            if(m->isArray) {
+                const size_t length = *((const size_t *) ptr);
+                ptr += sizeof(size_t);
+                ret = Array_encodeBinary(*(void *UA_RESTRICT const *) ptr, length, mt, ctx);
+                ptr += sizeof(void *);
+                continue;
+            }
+            /* Scalar */
+            ret = encodeWithExchangeBuffer((const void*)ptr, mt, ctx);
+
+            /* Array. Buffer-exchange is done inside Array_encodeBinary if required. */
+            ptr += mt->memSize;
+        }
+
+    }
+    ctx->depth--;
+    return ret;
+}
+
+static status
+encodeBinaryUnion(const void *src, const UA_DataType *type, Ctx *ctx) {
+    /* Check the recursion limit */
+    if(ctx->depth > UA_ENCODING_MAX_RECURSION)
+        return UA_STATUSCODE_BADENCODINGERROR;
+    ctx->depth++;
+
+    uintptr_t ptr = (uintptr_t)src;
+    status ret;
+    UA_UInt32 selection = *(UA_UInt32*) ptr;
+    if(selection == 0){
+        return encodeWithExchangeBuffer((const void*)ptr, &UA_TYPES[UA_TYPES_UINT32], ctx);
+    }
+    const UA_DataType *typelists[2] = { UA_TYPES, &type[-type->typeIndex] };
+    const UA_DataTypeMember *m = &type->members[selection-1];
+    const UA_DataType *mt = &typelists[!m->namespaceZero][m->memberTypeIndex];
+
+    ret = encodeWithExchangeBuffer((const void*)ptr, &UA_TYPES[UA_TYPES_UINT32], ctx);
```

Comment:
The code can be simplified so that encodeWithExchangeBuffer for the switchField occurs only once.